File: python/src/judge_jev/answers.py
"""Validation at the answer boundary shared by live, mock, and replay paths."""

from __future__ import annotations

import math
from typing import Any

from judge_jev.models import JudgeJevError, Rubric


def _number(value: Any, where: str) -> float:
    if isinstance(value, bool) or not isinstance(value, (int, float)):
        raise JudgeJevError(f"{where} must be a number, got {type(value).__name__}")
    result = float(value)
    if not math.isfinite(result):
        raise JudgeJevError(f"{where} must be finite")
    return result


def _unit(value: Any, where: str) -> float:
    result = _number(value, where)
    if not 0.0 <= result <= 1.0:
        raise JudgeJevError(f"{where} must be between 0 and 1")
    return result


def _probabilities(value: Any, where: str, allowed: set[str]) -> None:
    if not isinstance(value, dict):
        raise JudgeJevError(f"{where} must be an object")
    for key, probability in value.items():
        if not isinstance(key, str) or key not in allowed:
            raise JudgeJevError(
                f"{where} has invalid key {key!r}; expected one of {sorted(allowed)}"
            )
        _unit(probability, f"{where}.{key}")


def validate_answers(
    rubric: Rubric, answers: Any, *, require_any: bool = True
) -> dict[str, dict[str, Any]]:
    """Validate every present expected answer without inventing missing answers.

    Missing expected answers remain a routing concern and always escalate. A wholly
    empty response is an operational failure because no judgment was returned.
    """
    if not isinstance(answers, dict):
        raise JudgeJevError(
            f"system_one answers must be a JSON object, got {type(answers).__name__}"
        )
    if require_any and not answers:
        raise JudgeJevError("system_one returned no answers")

    unknown = sorted(set(answers) - set(rubric.questions))
    if unknown:
        raise JudgeJevError(f"answers contain unknown question ids: {', '.join(unknown)}")

    for answer_id, answer in answers.items():
        where = f"answers.{answer_id}"
        if not isinstance(answer, dict):
            raise JudgeJevError(f"{where} must be an object")
        question = rubric.questions[answer_id]
        expected = question["type"]
        actual = answer.get("type")
        if actual != expected:
            raise JudgeJevError(f"{where}.type must be {expected!r}, got {actual!r}")

        if expected == "noul":
            if "noul" not in answer:
                raise JudgeJevError(f"{where} is missing required field 'noul'")
            _unit(answer["noul"], f"{where}.noul")
            continue

        for field in ("confidence", "probabilities"):
            if field not in answer:
                raise JudgeJevError(f"{where} is missing required field {field!r}")
        _unit(answer["confidence"], f"{where}.confidence")

        criteria = question["criteria"]
        if expected == "choice":
            if "choice" not in answer:
                raise JudgeJevError(f"{where} is missing required field 'choice'")
            choice = answer["choice"]
            allowed = {str(label) for label in criteria}
            if not isinstance(choice, str) or choice not in allowed:
                raise JudgeJevError(
                    f"{where}.choice must be one of {sorted(allowed)}, got {choice!r}"
                )
            _probabilities(answer["probabilities"], f"{where}.probabilities", allowed)
        else:
            if "score" not in answer:
                raise JudgeJevError(f"{where} is missing required field 'score'")
            score = _number(answer["score"], f"{where}.score")
            highest = len(criteria) - 1
            if not 0.0 <= score <= highest:
                raise JudgeJevError(f"{where}.score must be between 0 and {highest}")
            allowed = {str(index) for index in range(len(criteria))}
            _probabilities(answer["probabilities"], f"{where}.probabilities", allowed)
            legend = answer.get("legend")
            if legend is not None:
                if not isinstance(legend, dict) or any(str(key) not in allowed for key in legend):
                    raise JudgeJevError(f"{where}.legend keys must be valid score levels")
    return answers

File: rust/src/answers.rs
//! Validation at the answer boundary shared by live, mock, and replay paths.

use crate::models::{Answer, Rubric};
use anyhow::{bail, Result};
use std::collections::{HashMap, HashSet};

fn unit(value: f64, where_: &str) -> Result<()> {
    if !value.is_finite() || !(0.0..=1.0).contains(&value) {
        bail!("{where_} must be finite and between 0 and 1");
    }
    Ok(())
}

fn probabilities(values: &HashMap<String, f64>, allowed: &HashSet<String>, where_: &str) -> Result<()> {
    for (key, probability) in values {
        if !allowed.contains(key) {
            let mut expected: Vec<&String> = allowed.iter().collect();
            expected.sort();
            bail!("{where_} has invalid key '{key}'; expected one of {expected:?}");
        }
        unit(*probability, &format!("{where_}.{key}"))?;
    }
    Ok(())
}

pub fn validate_answers(rubric: &Rubric, answers: &HashMap<String, Answer>) -> Result<()> {
    if answers.is_empty() {
        bail!("system_one returned no answers");
    }
    let mut unknown: Vec<&String> = answers
        .keys()
        .filter(|answer_id| !rubric.questions.contains_key(*answer_id))
        .collect();
    unknown.sort();
    if !unknown.is_empty() {
        bail!("answers contain unknown question ids: {unknown:?}");
    }

    for (answer_id, answer) in answers {
        let question = &rubric.questions[answer_id];
        let where_ = format!("answers.{answer_id}");
        if answer.kind() != question.qtype {
            bail!(
                "{where_}.type must be '{}', got '{}'",
                question.qtype,
                answer.kind()
            );
        }
        match answer {
            Answer::Noul { noul } => unit(*noul, &format!("{where_}.noul"))?,
            Answer::Choice {
                choice,
                confidence,
                probabilities: values,
            } => {
                unit(*confidence, &format!("{where_}.confidence"))?;
                let allowed: HashSet<String> = question
                    .criteria
                    .as_mapping()
                    .into_iter()
                    .flatten()
                    .filter_map(|(key, _)| key.as_str().map(str::to_string))
                    .collect();
                if !allowed.contains(choice) {
                    let mut expected: Vec<&String> = allowed.iter().collect();
                    expected.sort();
                    bail!("{where_}.choice must be one of {expected:?}, got '{choice}'");
                }
                probabilities(values, &allowed, &format!("{where_}.probabilities"))?;
            }
            Answer::Score {
                score,
                confidence,
                legend,
                probabilities: values,
            } => {
                unit(*confidence, &format!("{where_}.confidence"))?;
                let levels = question.criteria.as_sequence().map_or(0, Vec::len);
                if !score.is_finite() || *score < 0.0 || *score > (levels.saturating_sub(1)) as f64 {
                    bail!(
                        "{where_}.score must be finite and between 0 and {}",
                        levels.saturating_sub(1)
                    );
                }
                let allowed: HashSet<String> = (0..levels).map(|index| index.to_string()).collect();
                probabilities(values, &allowed, &format!("{where_}.probabilities"))?;
                if let Some(legend) = legend {
                    if legend.keys().any(|key| !allowed.contains(key)) {
                        bail!("{where_}.legend keys must be valid score levels");
                    }
                }
            }
        }
    }
    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::rubric::load_rubric;

    fn score(value: f64, confidence: f64, key: &str) -> Answer {
        Answer::Score {
            score: value,
            confidence,
            legend: None,
            probabilities: HashMap::from([(key.to_string(), 0.5)]),
        }
    }

    #[test]
    fn fractional_scores_are_valid_but_domains_are_enforced() {
        let rubric = load_rubric("assistant-reply").unwrap();
        let valid = HashMap::from([("score.helpfulness".to_string(), score(1.5, 0.5, "1"))]);
        validate_answers(&rubric, &valid).unwrap();
        for answer in [score(4.1, 0.5, "4"), score(1.5, -0.1, "1"), score(1.5, 0.5, "99")] {
            let bad = HashMap::from([("score.helpfulness".to_string(), answer)]);
            assert!(validate_answers(&rubric, &bad).is_err());
        }
    }

    #[test]
    fn noul_must_be_finite_and_in_unit_interval() {
        let rubric = load_rubric("assistant-reply").unwrap();
        for value in [f64::NAN, f64::INFINITY, -0.1, 1.1] {
            let bad = HashMap::from([("screen.judgeable".to_string(), Answer::Noul { noul: value })]);
            assert!(validate_answers(&rubric, &bad).is_err());
        }
    }
}
